Fix tauruseditor

cmft opened the pull request:
Remove backwards compatibility with spyder<3 and make TaurusEditor work in Taurus4

cpascual commented:
I did a quick code review on the commits since #a7cbe21 (I had already reviewed ones, I think) and found them  ok

cpascual commented:
Apart from whatever we find while testing, for the final pull we should consider rebasing (actually, squashing) this before merging to master

vxgmichel commented:
@cpascual Isn't `git merge --squash bug-187` enough? Is there any reason to keep intermediate commits?

cpascual commented:
@vxgmichel 
> @cpascual Isn't git merge --squash bug-187 enough?
I think it is enough

vxgmichel commented:
Then I think we're good!

NexeyaSGara commented:
Yeah good job everybody

cpascual commented:
Latest commits reviewed. They fix the exceptions I reported.

Here is one more exception: running `taurusgui example01`, with PyMca 4.7.4 installed to provide the nexus widget>

```
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/taurusgui/taurusgui.py", line 1158, in loadConfiguration
    w = p.getWidget(sdm=Qt.qApp.SDM, setModel=False)
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/taurusgui/utils.py", line 213, in getWidget
    w = klass()
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/extra_nexus/taurusnexuswidget.py", line 104, in __init__
    self.__fileModel.fileAppended.connect(self.treeWidget.fileAppended)
```

cpascual commented:
Another signal-related exception:

- launch `taurusplot eval:rand(12)`
- right-click on widget --> Plot configuration
- Change any of the curve properties (e.g. line style) and Apply:

```
Traceback (most recent call last):
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/plot/qwtdialog.py", line 418, in apply
    prop, names = self.curvesAppearanceChooser.onApply()
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/plot/curvesAppearanceChooserDlg.py", line 368, in onApply
    self.curveAppearanceChanged.emit(prop, names)
TypeError: CurvesAppearanceChooser.curveAppearanceChanged[tuple, list].emit(): argument 1 has unexpected type 'CurveAppearanceProperties'
```

cpascual commented:
Another exception in taurusplot:

- run `taurusplot 'eval:rand(12)'`
- right-click --> Save current Settings

```
Traceback (most recent call last):
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/plot/taurusplot.py", line 2460, in saveConfig
    ofile = open(ofile, 'w')
TypeError: coercing to Unicode: need string or buffer, bool found
```

Also, a similar one occurs when selecting "Retrieve Saved settings":
```
Traceback (most recent call last):
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/plot/taurusplot.py", line 2480, in loadConfig
    ifile = open(ifile, 'r')
TypeError: coercing to Unicode: need string or buffer, bool found

```

(tests to be continued next week...)

vxgmichel commented:
All fixed!

cpascual commented on lib/taurus/qt/qtgui/extra_nexus/taurusnexuswidget.py:
```diff
@@ -101,8 +101,8 @@ def __init__(self, *args, **kwargs):
         # connections
-        self.__fileModel.fileAppended.connect(self.treeWidget.fileAppended)
-        self.treeWidget.HDF5WidgetSignal.connect(self.onHDF5WidgetSignal)
+        self.__fileModel.sigFileAppended.connect(self.treeWidget.fileAppended)
```
This change does not seem to fix the issue. I now get 
```
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/taurusgui/taurusgui.py", line 1158, in loadConfiguration
    w = p.getWidget(sdm=Qt.qApp.SDM, setModel=False)
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/taurusgui/utils.py", line 213, in getWidget
    w = klass()
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/extra_nexus/taurusnexuswidget.py", line 104, in __init__
    self.__fileModel.sigFileAppended.connect(self.treeWidget.fileAppended)
AttributeError: 'FileModel' object has no attribute 'sigFileAppended'
```

cpascual commented:
Wow, that was fast!
You fix faster than I can report!

The fixes for the taurusplot exceptions work well for me. The fix for the pymca exception does not, though (see https://github.com/cpascual/taurus/pull/5#discussion-diff-64892987)

vxgmichel commented:
PyMca 4.7.4 doesn't provide new style signal. Is using PyMca5 ok?

cpascual commented on setup.py:
```diff
@@ -201,7 +201,7 @@ def get_release_info():
     'lxml (>=2.1)',  # [Taurus-TaurusGUI]
     'spyder (>=2.2)',  # [Taurus-Editor] --> or maybe move it to sardana
-    # Consider also guiqwt, guidata, PyMca,...
+    # Consider also guiqwt, guidata, PyMca5,...
```
This change in this file produces a merge conflict. Can you undo just this one?

cpascual commented on lib/taurus/qt/qtgui/extra_nexus/taurusnexuswidget.py:
```diff
@@ -101,8 +101,8 @@ def __init__(self, *args, **kwargs):
-        self.treeWidget.HDF5WidgetSignal.connect(self.onHDF5WidgetSignal)
+        self.__fileModel.sigFileAppended.connect(self.treeWidget.fileAppended)
+        self.treeWidget.sigHDF5WidgetSignal.connect(self.onHDF5WidgetSignal)
         self.openFileAction.triggered.connect(self.openFile)
```
The default overload of QAction.triggered uses a "bool" as argument. But we need the no-argument overload. So, `self.openFileAction.triggered[()].connect(self.openFile)` instead of `self.openFileAction.triggered.connect(self.openFile)`

cpascual commented:
error with taurusform compact mode:

- `taurusform sys/tg_test/1/long_spectrum`
- right-click --> Compact Mode

... And the compact widget fails to be created (shows "[Error]"). 
The following is printed in the debug log:

```
taurusform sys/tg_test/1/long_spectrum --taurus-log-level=trace
(...)
MainThread     DEBUG    2016-06-02 21:37:02,125 TaurusValue.sys/tg_test/1/long_spectrum: Error creating read widget:
'TaurusArrayEditorButton' object has no attribute 'applied'
```

cpascual commented:
Error in taurusplot export to ascii:

- Launch `taurusplot sys/tg_test/1/long_spectrum sys/tg_test/1/double_spectrum`
- right->click --> Export & Print --> Export To ASCII
- Change the selection combobox in the dialog: ==> nothing happens and exceptions are raised

```
Traceback (most recent call last):
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/panel/qdataexportdialog.py", line 151, in onDataSetCBChange
    self.updateText(key)
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/panel/qdataexportdialog.py", line 206, in updateText
    xdata, ydata = self.datadict[key]
KeyError: '0'

```

vxgmichel commented:
> 'TaurusArrayEditorButton' object has no attribute 'applied'

Here's the full traceback:

```
MainThread     INFO     2016-06-03 09:14:45,313 TaurusRootLogger: Using "PyQt4" for Qt
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/taurus/qt/qtgui/panel/taurusvalue.py", line 806, in updateReadWidget
    self._readWidget = self._newSubwidget(self._readWidget, klass)
  File "/usr/local/lib/python2.7/dist-packages/taurus/qt/qtgui/panel/taurusvalue.py", line 689, in _newSubwidget
    result = newClass()
  File "/usr/local/lib/python2.7/dist-packages/taurus/qt/qtgui/compact/abstractswitcher.py", line 145, in __init__
    self.setWriteWidget(self.writeWClass())
  File "/usr/local/lib/python2.7/dist-packages/taurus/qt/qtgui/compact/abstractswitcher.py", line 218, in setWriteWidget
    getattr(self.writeWidget, sig).connect(self.exitEdit)
AttributeError: 'TaurusArrayEditorButton' object has no attribute 'applied'
```

So it turns out `TaurusArrayEditorButton` is a write widget, even though it doesn't inherit from `TaurusBaseWritableWidget` (it does inherit from `TaurusBaseWidget` though). There are many ways to fix this. What do you recommend?

vxgmichel commented:
> Error in taurusplot export to ascii

Fixed!

cpascual commented:
Error in `taurustrend2d --demo`

...but note that I am testing with guiqwt 2.3.1 and maybe we should not worry now about guiqwt signal related errors, since we intend to move to guiqwt >=3 (which changed to new-style signals)

```
taurustrend2d --demo  
...
Traceback (most recent call last):
  File "/home/cpascual/.local/devenv/bin/taurustrend2d", line 9, in <module>
    load_entry_point('taurus', 'gui_scripts', 'taurustrend2d')()
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/extra_guiqwt/taurustrend2d.py", line 265, in taurusTrend2DMain
    buffersize=int(options.max_buffer_size))
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/extra_guiqwt/taurustrend2d.py", line 67, in __init__
    self.add_tool(toolklass)
  File "/home/cpascual/.local/lib/python2.7/site-packages/guiqwt/plot.py", line 277, in add_tool
    tool.register_plot(plot)
  File "/home/cpascual/src/taurus/lib/taurus/qt/qtgui/extra_guiqwt/tools.py", line 217, in register_plot
    baseplot.items_changed.connect(self.items_changed)
AttributeError: 'ImagePlot' object has no attribute 'items_changed'

```

vxgmichel commented:
Looks like `guiqwt` changed the [signal names](https://github.com/PierreRaybaut/guiqwt/blob/master/guiqwt/baseplot.py#L135):
```python
     SIG_ITEMS_CHANGED = Signal('PyQt_PyObject')
```
What should we do about it?

cpascual commented:
> Looks like guiqwt changed the signal names:
>      SIG_ITEMS_CHANGED = Signal('PyQt_PyObject')
> What should we do about it?

I think we should aim for supporting guiqwt >=3 and forget about guiqwt<3

We are going to do some tests here starting today, but It would be great if you could also test on your side and start fixing whatever you find in order to support it.

vxgmichel commented:
It seems like `item_changed` is the only `guiqwt`-specific signal:
```bash
lib/taurus/qt/qtgui/extra_guiqwt $ grep -R "\.connect("
taurustrend2d.py:        self.trendItem.dataChanged.connect(self.update_cross_sections)
curvesmodel.py:        self.ui.applyBT.clicked.connect(self.onApply)
curvesmodel.py:        self.ui.reloadBT.clicked.connect(self.onReload)
curvesmodel.py:        self.ui.cancelBT.clicked.connect(self.close)
curvesmodel.py:        self.ui.tangoTree.addModels.connect(self.onModelsAdded)
curvesmodel.py:        self.ui.curvesTable.customContextMenuRequested.connect(self.onTableContextMenu)
curvesmodel.py:        w.applied.connect(dlg.accept)
curvesmodel.py:        w.ui.cancelBT.clicked.connect(dlg.close)
plot.py:        self.imgItem.dataChanged.connect(self.update_cross_sections)
image.py:        qdoor.recordDataUpdated.connect(self.scanDataReceived)
image.py:        img.dataChanged.connect(win.update_cross_sections)
image.py:    # taurusimage.dataChanged.connect(win.update_cross_sections)
curve.py:        self.taurusEvent.connect(self.filterEvent)
tools.py:        baseplot.items_changed.connect(self.items_changed)
```

cpascual commented:
> It seems like item_changed is the only guiqwt-specific signal

You are right. I will make a PR now with a proposal for supporting guiqwt2 while simultaneously preparing for guiqwt3

vxgmichel commented:
Great!

cpascual commented:
I opened 3 PRs. If you accept them, IMHO the bug187 would be ready for squash-and-merge into develop.

Just for the record, here is a list of all the features that we manually tested (it is highly probable that we would have missed something, but  the basic things should be ok):

```

-Forms-

    -TaurusForm as a stand-alone application-
    -The widgets used for different types of attributes and devices-
    -Changing the contents of a form-
    -Drag and Drop support-
    -Interacting with the control system: writing to attributes-
    -Compact mode-

-Model Chooser-

-Plots-

    -TaurusPlot as a Stand-alone application-
    -Working with two Y scales--
    -TaurusPlot context menu-
    -Zooming and panning-
    -Plot Configuration dialog -
    -Choosing what is plotted-
    -Storing and recovering current configuration-
    -Obtaining information about the plotted values-
    -Exporting and printing the data --> !!! (reported)-
    -Customizing the titles of the curves-
    -Date/time support-

-TaurusTrend User’s Interface-

    -TaurusTrend as a stand-alone application-
    -Timestamp VS “event number” mode-
    -The Fixed-range X scale-
    -Using a spectrum as a collection of scalar values-
    Accessing archived values [PENDING]
    Polling buffer  [PENDING]
    -Forced read of attributes-
    -All context menu entries-

-Array Editor-

-Synoptics-

-TaurusPanel & TaurusDevicePanel-

    -TaurusDevicePanel as a stand-alone application-
    -TaurusPanel as a stand-alone application-


-TaurusGui-

    -Launching TaurusGUI-
    -General structure of a TaurusGUI application-
    -Re-arranging panels (moving, resizing, hiding,...)-
    -Creating custom panels-
    -Perspectives-
    -Synoptic panels-
    -External Application Launchers-
    -taurusgui --new-gui-
    -panel-instrument association-
    -export to xml-
    -export settings-
    -show shared data manager associations-
    -...-


-Configurations-

    -Taurus Configuration Browser Application-

-TaurusDemo-

    -Taurus Demo Application-
    -Containers-
    -button-
    -tree-
    -dialog-
    -table-
    -display-
    -panel-


-extra_guiqwt-

    -taurustrend2d-
    -taurustrend1d-
    -tauruscurve-
    -taurusimage-
    -autoscroll tool-
    -modelchooser tool-
    -axis tool-
    -zooms, contrast, cross-sections and other standard tools-

 
```

vxgmichel commented:
Are you setting some CI up? That's cool, what's the status of unit-testing in taurus by the way?

cpascual commented:
> Are you setting some CI up? That's cool, what's the status of unit-testing in taurus by the way? 

Yes. I set up for the taurus4-preview branch in my personal fork, as a way of evaluating Travis-CI. The reason you are seeing it now is that Yesterday I changed the travis.yml to trigger on the develop branch too.
Unfortunately, since we moved to setuptools, travis fails (we need to adapt it). I won't have time in 2 weeks, but maybe @cmft can fix the current issues

vxgmichel commented:
Good! The tests pass on my setup at least, except for some epic related stuff. 

Just a random thought: I noticed those tests require to have `sys/tg_test/1` running. Maybe this device could be started in a subprocess using the `--file` option? I guess it could be done quite easily by subclassing `TestCase`, overriding the `setUpClass` and `tearDownClass` methods and patching `TANGO_HOST`.

Also, +1 for the setuptools migration

cpascual commented:
> Maybe this device could be started in a subprocess using the --file option?

Yes, that is something we meant to do but never found time. By all means, feel free to contribute a patch (you know, now that this one is over...)  ;)

cmft commented:
In fact, taurus almost has the necessary tools to do it. In the tango
scheme test, we did something similar creating an instance of a device
server (start/test/stop) and deleting it at the end of the test.

2016-06-09 15:34 GMT+02:00 cpascual <notifications@github.com>:

> Maybe this device could be started in a subprocess using the --file option?
>
> Yes, that is something we meant to do but never found time. By all means,
> feel free to contribute a patch (you know, now that this one is over...) ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cpascual/taurus/pull/5#issuecomment-224895918>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AF4gkoamZ__-iSa8KCivKVPjj6wlUkVeks5qKBZxgaJpZM4IkX0A>
> .
>

vxgmichel commented:
One issue with the `--file` option is that it doesn't support events in tango 8. Does taurus also test event subscription?

cmft commented:
Not yet.  We have to implement this feature in the TaurusTangoScheme device
server, and also fix the attribute subscription in the tango scheme,
nowadays it  only take into account (the tango change event) .

2016-06-09 16:11 GMT+02:00 Vincent Michel <notifications@github.com>:

> One issue with the --file option is that it doesn't support events in
> tango 8. Does taurus also test event subscription?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cpascual/taurus/pull/5#issuecomment-224906555>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AF4gkn6dnlIdCNAKQ3pUeyy4lP0g4UIDks5qKB70gaJpZM4IkX0A>
> .
>

cmft committed:
Remove spyderlib imports

Delete spyderlib back. compatibility imports.
TaurusEditor only works with spyder in Taurus4.

cmft committed:
Adapt tauruseditor to spyder v3

TaurusEditor does not work in Taurus 4. Adapt it to work
with spyder V3.

cpascual commented on lib/taurus/qt/qtgui/editor/tauruseditor.py:
```diff
@@ -123,27 +115,32 @@ def reload(self, idx=None, filename=None, goto=None):
             editorstack.add_corner_widgets_to_tabbar([5, oe_btn])
 
-        editorstack.set_io_actions(*self.io_actions)
+        if len(self.io_actions) == 0:
```
Instead of this if, the createMenuActions method should return the dummy implementation (and document it in its docstring!)

cpascual commented on lib/taurus/qt/qtgui/editor/tauruseditor.py:
```diff
@@ -123,27 +115,32 @@ def reload(self, idx=None, filename=None, goto=None):
         self.editorStack().set_current_filename(filename)
 
+    def is_file_opened(self, filename=None):
+        # TODO: Implement it
```
This should not be a TODO, but a base implementation (documented as such).
I mean: the current immplementation is OK. It should not be marked with a TODO. Instead, the docstring should say something like: 

Dummy implementation that always return None. Reimplement in derived classes to return the index of already-open files in the editor_stack, or None if the file is not already open.

Commit:
Implement better dummy base createMenuActions

The existing createMenuActions returns empty lists, which forces some ugly handling
afterwards. Implement something usable (albeit dummy) instead, and document it.

Commit:
Properly document is_file_opened

is_file_opened is undocumentedand had a TODO tag but it actually 
provides a reasonable base implementation. Document it as such.
